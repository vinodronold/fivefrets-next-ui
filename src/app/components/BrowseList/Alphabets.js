import React from 'react'
import Link from 'next/link'
import { browse } from '../../constants/routes'
import Button from '../Button'
import Box from '../Box'

export default () => (
  <Box>
    {['0 - 1', ...'ABCDEFGHIJKLMNOPQRSTUVWXYZ'.split('')].map(a => (
      <Link prefetch key={a} href={browse.href(a.substr(0, 1))} as={browse.as(a.substr(0, 1))}>
        <Button style={{ padding: '.5rem' }}>{a}</Button>
      </Link>
    ))}
  </Box>
)
