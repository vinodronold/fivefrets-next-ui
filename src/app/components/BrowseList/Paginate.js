import React from 'react'
import { number, func } from 'prop-types'
import Button from '../Button'

const _propTypes = {
  curr_page: number,
  total_pages: number,
  MoveToBrowsePage: func.isRequired
}

const _defaultProps = {
  curr_page: 1,
  total_pages: 1
}

/// prop type validation is not happening in test

const Paginate = ({ curr_page, total_pages, MoveToBrowsePage }) => (
  <div>
    <Button
      disabled={curr_page === 1}
      onClick={() => {
        MoveToBrowsePage(curr_page - 1)
      }}>
      &#8606; Prev
    </Button>

    <Button
      onClick={() => {
        MoveToBrowsePage(1)
      }}>{`${curr_page} of ${total_pages}`}</Button>

    <Button
      disabled={curr_page === total_pages}
      onClick={() => {
        MoveToBrowsePage(curr_page + 1)
      }}>
      Next &#8608;
    </Button>

    <style jsx>{`
      div {
        padding: 1rem;
        display: flex;
        justify-content: center;
        flexwrap: wrap;
      }
    `}</style>
  </div>
)

Paginate.propTypes = _propTypes
Paginate.defaultProps = _defaultProps

export default Paginate
