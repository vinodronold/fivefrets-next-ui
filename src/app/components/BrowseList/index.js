import BrowseList from './BrowseList'
export default BrowseList
