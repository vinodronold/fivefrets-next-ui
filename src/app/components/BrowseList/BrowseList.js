import React, { PureComponent, Fragment } from 'react'
import { bool, number, func, arrayOf, object } from 'prop-types'
import Alphabets from './Alphabets'
import Paginate from './Paginate'
import SongList from '../SongList'

export default class BrowseList extends PureComponent {
  static propTypes = {
    isError: bool,
    curr_page: number,
    total_pages: number,
    songs: arrayOf(object),
    MoveToBrowsePage: func,
    url: object
  }

  static defaultProps = {
    isError: false,
    curr_page: 1,
    total_pages: 1,
    url: {
      start: 'a'
    },
    songs: []
  }

  handleSongsFetch(start, curr_page, songs) {
    if (songs.length === 0) {
      this.props.FetchBrowsePageSongs(start, curr_page)
    }
  }

  componentWillMount() {
    const { url, curr_page, songs } = this.props
    this.handleSongsFetch(url.start, curr_page, songs)
  }

  componentWillReceiveProps(nextProps) {
    const { url, curr_page, songs } = nextProps
    if (url.start !== this.props.url.start) {
      this.props.FetchBrowsePageSongs(url.start, 1)
    }
    if (curr_page !== this.props.curr_page) {
      this.handleSongsFetch(url.start, curr_page, songs)
    }
  }

  render() {
    const { isError, curr_page, total_pages, songs, MoveToBrowsePage } = this.props
    return (
      <Fragment>
        <Alphabets />
        <SongList isError={isError} songs={songs} />
        <Paginate curr_page={curr_page} total_pages={total_pages} MoveToBrowsePage={MoveToBrowsePage} />
      </Fragment>
    )
  }
}
