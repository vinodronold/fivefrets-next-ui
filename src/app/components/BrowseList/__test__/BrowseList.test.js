import React from 'react'
import renderer from 'react-test-renderer'
import { mount } from 'enzyme'
import { spy } from 'sinon'
import BrowseList from '../BrowseList'

jest.mock('next/router')

describe('components/BrowseList', () => {
  it('should match snapshot', () => {
    const _browse_list = renderer.create(<BrowseList FetchBrowsePageSongs={jest.fn()} />)
    expect(_browse_list).toMatchSnapshot()
  })

  it('componentWillReceiveProps - for page number change', () => {
    const _spy_componentWillReceiveProps = spy(BrowseList.prototype, 'componentWillReceiveProps')
    const _browse_list = mount(<BrowseList FetchBrowsePageSongs={jest.fn()} />)
    expect(_spy_componentWillReceiveProps.calledOnce).toEqual(false)
    _browse_list.setProps({ curr_page: 2 })
    expect(_spy_componentWillReceiveProps.calledOnce).toEqual(true)
    _spy_componentWillReceiveProps.restore()
  })

  it('componentWillReceiveProps - browse start change', () => {
    const _spy_componentWillReceiveProps = spy(BrowseList.prototype, 'componentWillReceiveProps')
    const _browse_list = mount(<BrowseList FetchBrowsePageSongs={jest.fn()} />)
    expect(_spy_componentWillReceiveProps.calledOnce).toEqual(false)
    _browse_list.setProps({ url: { start: '*' } })
    expect(_spy_componentWillReceiveProps.calledOnce).toEqual(true)
    _spy_componentWillReceiveProps.restore()
  })

  it('handleSongsFetch should be called when mounting and page change', () => {
    const _spy_handleSongsFetch = spy(BrowseList.prototype, 'handleSongsFetch')
    const _browse_list = mount(<BrowseList FetchBrowsePageSongs={jest.fn()} />)
    expect(_spy_handleSongsFetch.calledOnce).toEqual(true)
    _browse_list.setProps({ curr_page: 2 })
    expect(_spy_handleSongsFetch.calledTwice).toEqual(true)
    expect(_spy_handleSongsFetch.callCount).toEqual(2)
    _spy_handleSongsFetch.restore()
  })

  it('FetchBrowsePageSongs should be called when Songs not availble', () => {
    const _spy_FetchBrowsePageSongs = spy()
    mount(<BrowseList FetchBrowsePageSongs={_spy_FetchBrowsePageSongs} />)
    expect(_spy_FetchBrowsePageSongs.called).toEqual(true)
  })

  it('FetchBrowsePageSongs should not be called when Songs availble', () => {
    const _spy_FetchBrowsePageSongs = spy()
    mount(<BrowseList FetchBrowsePageSongs={_spy_FetchBrowsePageSongs} songs={[{ songs: 'test' }]} />)
    expect(_spy_FetchBrowsePageSongs.notCalled).toEqual(true)
  })
})
