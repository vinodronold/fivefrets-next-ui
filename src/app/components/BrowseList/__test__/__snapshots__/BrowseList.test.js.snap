// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/BrowseList should match snapshot 1`] = `
Array [
  <div
    className="jsx-3971180074 "
  >
    <button
      className="jsx-67710159 jsx-4255320455 "
      disabled={false}
      onClick={[Function]}
      style={
        Object {
          "padding": ".5rem",
        }
      }
    >
      0 - 1
    </button>
    <button
      className="jsx-67710159 jsx-4255320455 "
      disabled={false}
      onClick={[Function]}
      style={
        Object {
          "padding": ".5rem",
        }
      }
    >
      A
    </button>
    <button
      className="jsx-67710159 jsx-4255320455 "
      disabled={false}
      onClick={[Function]}
      style={
        Object {
          "padding": ".5rem",
        }
      }
    >
      B
    </button>
    <button
      className="jsx-67710159 jsx-4255320455 "
      disabled={false}
      onClick={[Function]}
      style={
        Object {
          "padding": ".5rem",
        }
      }
    >
      C
    </button>
    <button
      className="jsx-67710159 jsx-4255320455 "
      disabled={false}
      onClick={[Function]}
      style={
        Object {
          "padding": ".5rem",
        }
      }
    >
      D
    </button>
    <button
      className="jsx-67710159 jsx-4255320455 "
      disabled={false}
      onClick={[Function]}
      style={
        Object {
          "padding": ".5rem",
        }
      }
    >
      E
    </button>
    <button
      className="jsx-67710159 jsx-4255320455 "
      disabled={false}
      onClick={[Function]}
      style={
        Object {
          "padding": ".5rem",
        }
      }
    >
      F
    </button>
    <button
      className="jsx-67710159 jsx-4255320455 "
      disabled={false}
      onClick={[Function]}
      style={
        Object {
          "padding": ".5rem",
        }
      }
    >
      G
    </button>
    <button
      className="jsx-67710159 jsx-4255320455 "
      disabled={false}
      onClick={[Function]}
      style={
        Object {
          "padding": ".5rem",
        }
      }
    >
      H
    </button>
    <button
      className="jsx-67710159 jsx-4255320455 "
      disabled={false}
      onClick={[Function]}
      style={
        Object {
          "padding": ".5rem",
        }
      }
    >
      I
    </button>
    <button
      className="jsx-67710159 jsx-4255320455 "
      disabled={false}
      onClick={[Function]}
      style={
        Object {
          "padding": ".5rem",
        }
      }
    >
      J
    </button>
    <button
      className="jsx-67710159 jsx-4255320455 "
      disabled={false}
      onClick={[Function]}
      style={
        Object {
          "padding": ".5rem",
        }
      }
    >
      K
    </button>
    <button
      className="jsx-67710159 jsx-4255320455 "
      disabled={false}
      onClick={[Function]}
      style={
        Object {
          "padding": ".5rem",
        }
      }
    >
      L
    </button>
    <button
      className="jsx-67710159 jsx-4255320455 "
      disabled={false}
      onClick={[Function]}
      style={
        Object {
          "padding": ".5rem",
        }
      }
    >
      M
    </button>
    <button
      className="jsx-67710159 jsx-4255320455 "
      disabled={false}
      onClick={[Function]}
      style={
        Object {
          "padding": ".5rem",
        }
      }
    >
      N
    </button>
    <button
      className="jsx-67710159 jsx-4255320455 "
      disabled={false}
      onClick={[Function]}
      style={
        Object {
          "padding": ".5rem",
        }
      }
    >
      O
    </button>
    <button
      className="jsx-67710159 jsx-4255320455 "
      disabled={false}
      onClick={[Function]}
      style={
        Object {
          "padding": ".5rem",
        }
      }
    >
      P
    </button>
    <button
      className="jsx-67710159 jsx-4255320455 "
      disabled={false}
      onClick={[Function]}
      style={
        Object {
          "padding": ".5rem",
        }
      }
    >
      Q
    </button>
    <button
      className="jsx-67710159 jsx-4255320455 "
      disabled={false}
      onClick={[Function]}
      style={
        Object {
          "padding": ".5rem",
        }
      }
    >
      R
    </button>
    <button
      className="jsx-67710159 jsx-4255320455 "
      disabled={false}
      onClick={[Function]}
      style={
        Object {
          "padding": ".5rem",
        }
      }
    >
      S
    </button>
    <button
      className="jsx-67710159 jsx-4255320455 "
      disabled={false}
      onClick={[Function]}
      style={
        Object {
          "padding": ".5rem",
        }
      }
    >
      T
    </button>
    <button
      className="jsx-67710159 jsx-4255320455 "
      disabled={false}
      onClick={[Function]}
      style={
        Object {
          "padding": ".5rem",
        }
      }
    >
      U
    </button>
    <button
      className="jsx-67710159 jsx-4255320455 "
      disabled={false}
      onClick={[Function]}
      style={
        Object {
          "padding": ".5rem",
        }
      }
    >
      V
    </button>
    <button
      className="jsx-67710159 jsx-4255320455 "
      disabled={false}
      onClick={[Function]}
      style={
        Object {
          "padding": ".5rem",
        }
      }
    >
      W
    </button>
    <button
      className="jsx-67710159 jsx-4255320455 "
      disabled={false}
      onClick={[Function]}
      style={
        Object {
          "padding": ".5rem",
        }
      }
    >
      X
    </button>
    <button
      className="jsx-67710159 jsx-4255320455 "
      disabled={false}
      onClick={[Function]}
      style={
        Object {
          "padding": ".5rem",
        }
      }
    >
      Y
    </button>
    <button
      className="jsx-67710159 jsx-4255320455 "
      disabled={false}
      onClick={[Function]}
      style={
        Object {
          "padding": ".5rem",
        }
      }
    >
      Z
    </button>
  </div>,
  <div
    className="jsx-3649670468 container"
  >
    <div
      className="jsx-3649670468 loader"
    >
      <div
        className="jsx-791847780 jsx-2291006144 bar"
      />
      <div
        className="jsx-791847780 jsx-3500035231 bar"
      />
      <div
        className="jsx-791847780 jsx-1901401244 bar"
      />
      <div
        className="jsx-791847780 jsx-40336921 bar"
      />
      <div
        className="jsx-791847780 jsx-3370661658 bar"
      />
    </div>
    <div
      className="jsx-4247411757 jsx-3754337244"
      style={Object {}}
    >
      LOADING...
    </div>
  </div>,
  <div
    className="jsx-2366590580"
  >
    <button
      className="jsx-67710159 jsx-3220588738 "
      disabled={true}
      onClick={[Function]}
    >
      ↞ Prev
    </button>
    <button
      className="jsx-67710159 jsx-4255320455 "
      disabled={false}
      onClick={[Function]}
    >
      1 of 1
    </button>
    <button
      className="jsx-67710159 jsx-3220588738 "
      disabled={true}
      onClick={[Function]}
    >
      Next ↠
    </button>
  </div>,
]
`;
